topdown/eval.go

```diff
@@ -2332,34 +2367,38 @@ func (e evalVirtualPartial) evalTerm(iter unifyIterator, term *ast.Term, termbin
 	return eval.eval(iter)
 }
 
-func (e evalVirtualPartial) evalCache(iter unifyIterator) (ast.Ref, bool, error) {
+func (e evalVirtualPartial) evalCache(iter unifyIterator) (evalVirtualPartialCacheHint, error) {
+
+	var hint evalVirtualPartialCacheHint
 
 	if e.e.unknown(e.ref[:e.pos+1], e.bindings) {
-		return nil, false, nil
+		return hint, nil
 	}
 
-	var cacheKey ast.Ref
-
-	if e.ir.Kind == ast.PartialObjectDoc {
-
-		plugged := e.bindings.Plug(e.ref[e.pos+1])
+	if cached := e.e.virtualCache.Get(e.plugged[:e.pos+1]); cached != nil { // have full extent cached
+		e.e.instr.counterIncr(evalOpVirtualCacheHit)
+		hint.hit = true
+		return hint, e.evalTerm(iter, e.pos+1, cached, e.bindings)
+	}
 
-		if plugged.IsGround() {
-			path := e.plugged[:e.pos+2]
-			path[len(path)-1] = plugged
-			cached := e.e.virtualCache.Get(path)
+	plugged := e.bindings.Plug(e.ref[e.pos+1])
 
-			if cached != nil {
-				e.e.instr.counterIncr(evalOpVirtualCacheHit)
-				return nil, true, e.evalTerm(iter, cached, e.bindings)
-			}
+	if plugged.IsGround() {
+		hint.key = append(e.plugged[:e.pos+1], plugged)
 
-			e.e.instr.counterIncr(evalOpVirtualCacheMiss)
-			cacheKey = path
+		if cached := e.e.virtualCache.Get(hint.key); cached != nil {
+			e.e.instr.counterIncr(evalOpVirtualCacheHit)
+			hint.hit = true
+			return hint, e.evalTerm(iter, e.pos+2, cached, e.bindings)
 		}
+	} else {
+		hint.full = true
```

Comment:
One of the nice features of topdown is that we can push down partially ground terms into child queries. This is somewhat contrived but it demonstrates what I'm saying...

```rego
p {
  q[["foo", x]]
}

q[[a,b]] {
   r[a] = b
}

r := {"foo": 1, "bar": 2, "baz": 3}
```

Prior to this change, `a` would be ground when evaluating `r[a] = b` and so the evaluator would perform a constant-time lookup on `r`. With this change, that's no longer the case. The evaluator will have to iterate on `r[a] = b` and so the runtime complexity of `p` would no longer be constant...

I'm not sure how widely this feature is used but this could cause a regression. If possible, I think it would be better to only force full evaluation when the term is an unbound variable.

If you make this change, it should affect this test case in the PR: `partial set: each rule (non-ground composite), full extent cached`.

Reply:
Thanks, I wasn't aware of the full picture 💡 Updated!